File: internal/installer/gitclone/gitclone.go
// Package gitclone implements the "git-clone" step type.
package gitclone

import (
	"context"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"gearup/internal/config"
	gearexec "gearup/internal/exec"
)

// Installer is the git-clone step installer.
type Installer struct {
	Runner gearexec.Runner
}

// Check reports whether the destination directory already exists.
func (i *Installer) Check(_ context.Context, step config.Step) (bool, error) {
	if step.Dest == "" {
		return false, fmt.Errorf("git-clone step %q missing dest", step.Name)
	}
	dest := expandHome(step.Dest)
	info, err := os.Stat(dest)
	if err != nil {
		if os.IsNotExist(err) {
			return false, nil
		}
		return false, err
	}
	return info.IsDir(), nil
}

// Install runs `git clone [--branch <ref>] <repo> <dest>`.
func (i *Installer) Install(ctx context.Context, step config.Step) error {
	if step.Repo == "" {
		return fmt.Errorf("git-clone step %q missing repo", step.Name)
	}
	if step.Dest == "" {
		return fmt.Errorf("git-clone step %q missing dest", step.Name)
	}
	dest := expandHome(step.Dest)

	// Ensure parent directory exists.
	if err := os.MkdirAll(filepath.Dir(dest), 0o755); err != nil {
		return fmt.Errorf("git-clone %s: create parent dir: %w", step.Name, err)
	}

	cmd := fmt.Sprintf("git clone %s %s", step.Repo, dest)
	if step.Ref != "" {
		cmd = fmt.Sprintf("git clone --branch %s %s %s", step.Ref, step.Repo, dest)
	}
	res, err := i.Runner.RunQuiet(ctx, cmd)
	if err != nil {
		return err
	}
	if res.ExitCode != 0 {
		return fmt.Errorf("git clone %s failed (exit %d): %s", step.Name, res.ExitCode, res.Stderr)
	}
	return nil
}

func expandHome(p string) string {
	if strings.HasPrefix(p, "~/") {
		home, err := os.UserHomeDir()
		if err == nil {
			return filepath.Join(home, p[2:])
		}
	}
	return p
}

File: src/steps/git-clone.ts
import type { GitCloneStep } from "../schema"
import type { Context } from "../context"
import type { CheckResult, InstallResult } from "./types"
import path from "node:path"

function expandHome(p: string, home: string): string {
  if (p === "~") return home
  if (p.startsWith("~/")) return `${home}/${p.slice(2)}`
  return p
}

export async function checkGitClone(step: GitCloneStep, ctx: Context): Promise<CheckResult> {
  if (step.check) {
    const result = await ctx.exec.run({ argv: [step.check], shell: true })
    return result.exitCode === 0 ? { installed: true } : { installed: false }
  }

  const home = ctx.env.HOME ?? ""
  const dest = expandHome(step.dest, home)
  const result = await ctx.exec.run({ argv: ["test", "-d", dest] })
  return result.exitCode === 0 ? { installed: true } : { installed: false }
}

export async function installGitClone(step: GitCloneStep, ctx: Context): Promise<InstallResult> {
  const home = ctx.env.HOME ?? ""
  const dest = expandHome(step.dest, home)
  const parent = path.dirname(dest)

  const mkdirResult = await ctx.exec.run({ argv: ["mkdir", "-p", parent] })
  if (mkdirResult.exitCode !== 0) {
    return {
      ok: false,
      error: `git-clone ${step.name}: create parent dir ${parent} failed (exit ${mkdirResult.exitCode}): ${mkdirResult.stderr.trim()}`,
    }
  }

  const argv = step.ref
    ? ["git", "clone", "--branch", step.ref, step.repo, dest]
    : ["git", "clone", step.repo, dest]

  const cloneResult = await ctx.exec.run({ argv })
  if (cloneResult.exitCode === 0) return { ok: true }
  return {
    ok: false,
    error: `git clone ${step.name} failed (exit ${cloneResult.exitCode}): ${cloneResult.stderr.trim()}`,
  }
}
